feat(vim.fs): find(), dir() can "follow" symlinks

mike325 opened the pull request:
This PR introduce a new option for `vim.fs.dir` and `vim.fs.find` to follow symbolic links.
The default behaviour now turns on the `follow` feature, users can opt out of it with `follow=false`

mike325 committed:
feat: add follow opt to vim.fs.find

mike325 committed:
feat: add follow opt for vim.fs.dir

dundargoc commented:
nitpick: I think there's a risk of infinite recursion if the symbolic link is circular. To handle this, vimscript's `resolve()` just aborts after 100 times

> To cope with link cycles, resolving of symbolic links is stopped after 100 iterations.

I think it should also be possible to detect if it's a circular loop and abort in that case (I have vague memories of some path libraries going this route, but not swearing on it).


For what it's worth I think the PR is fine as is currently, mentioning it mostly for awareness or if it's an edge case we want to handle in the future.

mike325 committed:
fix: vim.fs.dir test

Force follow arg value by passing it to exec_lua function
Use absolute path to create link

mike325 commented:
> nitpick: I think there's a risk of infinite recursion if the symbolic link is circular. Vimscript's `resolve()` handles this with the following behavior:
> 
> > To cope with link cycles, resolving of symbolic links is stopped after 100 iterations.
> 
> I think it should also be possible to detect if it's a circular loop and abort in that case (I have vague memories of some path libraries going this route, but not swearing on it).
> 
> For what it's worth I think the PR is fine as is currently, mentioning it mostly for awareness or if it's an edge case we want to handle in the future.

I also thought about the infinite recursion problem since I myself encountered this issue in a repo I was working on, although I found it using `fd|rg`.

dundargoc commented:
I tested out this PR, there is some already some internal mechanism that prevents infinite recursion but I don't know what. I can "only" recurse 33 times before it aborts: `/Users/dundargoc/programs/neovim/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link/link`

justinmk commented on test/functional/lua/fs_spec.lua:
```diff
@@ -213,17 +214,59 @@ describe('vim.fs', function()
       )
       eq({ nvim_prog }, vim.fs.find(nvim_prog_basename, { path = test_build_dir, type = 'file' }))
 
+      eq(
+        { nvim_prog, test_source_path .. '/build_link/bin/' .. nvim_prog_basename },
+        vim.fs.find(nvim_prog_basename, {
+          path = test_source_path,
+          type = 'file',
+          limit = 2,
+          follow = true,
+        })
+      )
+
+      eq(
+        { nvim_prog },
```
there's already a `describe('find()'` block, and no setup in this `it()`, so it makes sense to move these 2 `eq()` assertions into a separate `it('follow=true'` block. That makes it clear what these are testing, at zero code/verbosity-cost.

mike325 committed:
fix: extract find follow symlink test

mike325 commented:
Should this change be included in the `new.txt` file ? 
Also, is it ok to have it `follow=true`? or should it keep the old behavior by using  `follow=false` by default instead?

justinmk commented:
is there a test for the infinite loop case?

> Should this change be included in the `new.txt` file ?

definitely. probably the "Lua" section of news-features.

justinmk commented:
> Also, is it ok to have it `follow=true`? or should it keep the old behavior by using `follow=false` by default instead?

if we have a test for the infinite-loop case then defaulting to true seems fine.

justinmk commented:
> a test for the infinite-loop case then defaulting to true seems fine.
> 
> should be a separate test case like `it('follow=true handles symlink loop')`

the test can be skipped on windows for now, if that makes it easier

mike325 committed:
fix: add link loop test and update news.txt

justinmk commented on test/testutil.lua:
```diff
@@ -394,15 +394,18 @@ end
   return not not (
     (s == 'win' and (sysname:find('windows') or sysname:find('mingw')))
     or (s == 'mac' and sysname == 'darwin')
+    or (s == 'linux' and sysname == 'linux')
```
can name it `unix`, and just assume that if nothing else matches then we treat it as `unix`

justinmk commented on test/functional/lua/fs_spec.lua:
```diff
@@ -166,8 +179,8 @@ describe('vim.fs', function()
-      local function run(dir, depth, skip)
-        return exec_lua(function()
+      local function run(dir, depth, skip, follow)
+        return exec_lua(function(follow_sym)
```
nit: usual convention is something like:
```suggestion
        return exec_lua(function(follow_)
```

justinmk commented on test/functional/lua/fs_spec.lua:
```diff
@@ -17,6 +17,19 @@ local mkdir = t.mkdir
 
 local nvim_prog_basename = is_os('win') and 'nvim.exe' or 'nvim'
 
+local platfom_link_limit = {
+  linux = 41,
+  mac = 33,
+  win = 64,
+}
+
+local link_limit = platfom_link_limit.linux
+if is_os('win') then
+  link_limit = platfom_link_limit.win
+elseif is_os('mac') or is_os('bsd') then
+  link_limit = platfom_link_limit.mac
+end
```
```suggestion
local link_limit = is_os('win') and 64
    or (is_os('mac') or is_os('bsd')) and 33 or 41
```

justinmk commented on test/testutil.lua:
```diff
@@ -405,10 +408,10 @@ end
 --- @return boolean
 function M.is_arch(s)
-  if not (s == 'x86_64' or s == 'arm64') then
+  if not (s == 'x86_64' or s == 'arm64' or s == 'aarch64') then
```
why is "aarch64" needed? not used in this PR.

and generally, we should have one token that means "arm", and then `is_arch` (or whatever calls it) should do the right thing for each platform.

mike325 replied:
oooo sorry about that, I was testing the linux side using a raspberry pi and a desktop, I wanted to know if the limit was different between x86 and ARM.
linux reports `aarch64` not `arm64` (or at least reaspbian does), I can remove it since it turned out not to be needed.

justinmk replied:
ok let's remove it?

justinmk commented:
LGTM, waiting for `is_arch` change to be dropped

justinmk commented:
> > Also, is it ok to have it `follow=true`? or should it keep the old behavior by using `follow=false` by default instead?
> 
> if we have a test for the infinite-loop case then defaulting to true seems fine.

We probably need the default to be `follow=false`. There are some reports of "performance regressions" from `follow=true`.

mike325 commented:
> We probably need the default to be `follow=false`. There are some reports of "performance regressions" from `follow=true`.
Do you want me to create an issue to track this @justinmk  ?

justinmk commented:
@mike325 i would prefer a PR :) we are release 0.11 soon, so it's urgent
